### PR TITLE
Moved discovery service installation script into the identity script.

### PR DESCRIPTION
**Pull request by hc-jordenlowe**
right now, everything is installed altogether.  Perhaps I should remove IDPSS and Discovery, then add them to a script separately from identity.  That script would call all 3

### CODE REVIEW

**Comment by hc-jordenlowe** [comment]
The new Interface of this install script is: 

- Install-Identity-Discovery-IdPSS.ps1  and it will have the same switches as Install-Identity.ps1.

There will be 2 config values being added to this:

- In discovery service scope, enableOAuth which will install identity first and allow discovery service to talk OAuth to everyone
- In Identity scope, useAzureAD will allow Azure AD to be used.

File: Fabric.Identity.API/scripts/Install-Discovery.ps1 [hunk 1]
```diff
@@ -0,0 +1,76 @@
+﻿param(
+    [PSCredential] $credential, 
+    [Hashtable] $configStore = @{Type = "File"; Format = "XML"; Path = "$PSScriptRoot\install.config"},
+    [switch] $quiet
+)
+
+Import-Module -Name .\Install-Discovery-Utilities.psm1 -Force
```
Comment (ctediz) [inline Fabric.Identity.API/scripts/Install-Discovery.ps1]: Is the discovery utilities not being brought over as well?

File: Fabric.Identity.API/scripts/Install-Discovery.ps1 [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+if([string]::IsNullOrEmpty($commonConfig.clientEnvironment))
+{
+  Write-DosMessage -Level "Error" -Message "It is required to have 'clientEnvironment' populated in install.config."
+}
+
+$discoveryIisUser = Get-IISAppPoolUser -credential $credential -appName $discoveryConfig.appName -storedIisUser $discoveryConfig.iisUser -installConfigPath $configStore.Path -scope $discoverySettingsScope
+
+$discoveryStandalonePath = ".\Catalyst.DiscoveryService.zip"
+$discoveryInstallerPath = "..\WebDeployPackages\Catalyst.DiscoveryService.zip"
```
Comment (ctediz) [inline Fabric.Identity.API/scripts/Install-Discovery.ps1]: Phew I forgot that was in there, I was worried about the discovery zip location

File: Fabric.Identity.API/scripts/Install-Identity-Discovery-IdPSS.ps1 [hunk 3]
```diff
@@ -0,0 +1,62 @@
+﻿#Requires -RunAsAdministrator
+#Requires -Version 5.1
+#Requires -Modules PowerShellGet, PackageManagement
+
+param(
+    [PSCredential] $credential,
```
Comment (ctediz) [inline Fabric.Identity.API/scripts/Install-Identity-Discovery-IdPSS.ps1]: Should there be multiple options for credentials now? One for Discovery and for Identity/IdPSS?

File: Fabric.Identity.API/scripts/Install-Identity-Discovery-IdPSS.ps1 [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+$discoveryConfig = Get-DosConfigValues -ConfigStore $configStore -Scope $discoveryScope
+$enableOAuth = [string]::IsNullOrEmpty($discoveryConfig.enableOAuth) -ne $true -and $discoveryConfig.enableOAuth -eq "true"
+
+if(!$enableOAuth){
+    # If AAD is not enabled, install discovery first.
+    .\Install-Discovery.ps1 -credential $credential -configStore $configStore -quiet
+}
+
+# Call the Identity powershell script
+.\Install-Identity.ps1 -credential $credential -configStore $configStore -noDiscoveryService:$noDiscoveryService -quiet
```
Comment (ctediz) [inline Fabric.Identity.API/scripts/Install-Identity-Discovery-IdPSS.ps1]: Identity/IdPSS are not using a config store, they're using the path to the install.config instead

File: Fabric.Identity.API/scripts/Install-Identity-Utilities.psm1 [hunk 5]
```diff
@@ -1406,6 +1406,116 @@ function Get-CurrentUserDomain
...
+        [Parameter(Mandatory=$true)]
+        [Hashtable] $discoveryConfig,
+        [Parameter(Mandatory=$true)]
+        [Hashtable] $commonConfig,
+        [Parameter(Mandatory=$true)]
+        [string] $userName,
+        [string] $registrationApiSecret
+    )
+
+    Confirm-DiscoveryConfig -discoveryConfig $discoveryConfig -commonConfig $commonConfig
```
Comment (ctediz) [inline Fabric.Identity.API/scripts/Install-Identity-Utilities.psm1]: If Confirm-DiscoveryConfig in the Discovery-Utilities module?

**Comment by ctediz** [comment]
One more general comment, if the Discovery install scripts are coming into the identity repo, does the identity install.config need to also have the values from the discovery install.config as well?

File: Fabric.Identity.API/scripts/Install-Discovery.ps1 (hunk 1 above)
Comment (hc-jordenlowe) [reply Fabric.Identity.API/scripts/Install-Discovery.ps1]: I have merged everything into the identity utilities.

File: Fabric.Identity.API/scripts/Install-Identity-Discovery-IdPSS.ps1 (hunk 4 above)
Comment (hc-jordenlowe) [reply Fabric.Identity.API/scripts/Install-Identity-Discovery-IdPSS.ps1]: So the script had it like this already.  I am pretty sure it uses the config store if it is presented.  This case is when Identity is installing.  If seperately, then the scenario you talk about is right.

File: Fabric.Identity.API/scripts/Install-Identity-Utilities.psm1 (hunk 5 above)
Comment (hc-jordenlowe) [reply Fabric.Identity.API/scripts/Install-Identity-Utilities.psm1]: Forgot to move this over

**Comment by hc-jordenlowe** [comment]
> One more general comment, if the Discovery install scripts are coming into the identity repo, does the identity install.config need to also have the values from the discovery install.config as well?

They do.

**Comment by hc-jordenlowe** [comment]
> Yah. Having the ability to run the scripts separately and then a master script to combine them makes sense. We have tests for Identity-Utilities and Idpss-Utilities, can we bring Discovery-Utilities tests here as well.

All of the tests are merged into identity's tests.

File: Fabric.Identity.API/scripts/Install-Identity-Discovery-IdPSS.ps1 (hunk 3 above)
Comment (hc-jordenlowe) [reply Fabric.Identity.API/scripts/Install-Identity-Discovery-IdPSS.ps1]: i see what you mean, ill fix that

File: Fabric.Identity.API/scripts/Install-Identity-Utilities.psm1 [hunk 6]
```diff
@@ -1406,6 +1436,133 @@ function Get-CurrentUserDomain
...
+        $webDeployParameters += @{
+            Name = "Fabric.Identity URL";
+            Value = "$($commonConfig.IdentityService)"
+        }
+    }
+    
+    return $webDeployParameters
+}
+
+function Confirm-DiscoveryConfig{
```
Comment (ctediz) [inline Fabric.Identity.API/scripts/Install-Identity-Utilities.psm1]: At some point we should probably create something similar for identity/idpss and have that run in the parent coordination script so the script fails early before trying to install with bad values/quiet mode

File: Fabric.Identity.API/scripts/Install-Identity-Discovery-IdPSS.ps1 [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+    [ValidateScript({
+        if (!(Test-Path $_)) {
+            throw "Path $_ does not exist. Please enter valid path to the install.config."
+        }
+        if (!(Test-Path $_ -PathType Leaf)) {
+            throw "Path $_ is not a file. Please enter a valid path to the install.config."
+        }
+        return $true
+    })] 
+    [string] $installConfigPath = "$PSScriptRoot\install.config",
```
[the remark this answers is not in the document]
Comment (hc-jordenlowe) [reply Fabric.Identity.API/scripts/Install-Identity-Discovery-IdPSS.ps1]: This is the interface to the script.  If someone wants to run each individual script, they can use the config store.  Thats at least following the path that was laid out before.

File: Fabric.Identity.API/scripts/Install-Identity-Discovery-IdPSS.ps1 [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+# Import Fabric Install Utilities
+$fabricInstallUtilities = ".\Fabric-Install-Utilities.psm1"
+if (!(Test-Path $fabricInstallUtilities -PathType Leaf)) {
+    Write-DosMessage -Level "Warning" -Message "Could not find fabric install utilities. Manually downloading and installing"
+    Get-WebRequestDownload -Uri https://raw.githubusercontent.com/HealthCatalyst/InstallScripts/master/common/Fabric-Install-Utilities.psm1 -NoCache -OutFile $fabricInstallUtilities
+}
+Import-Module -Name $fabricInstallUtilities -Force
+Test-MeetsMinimumRequiredPowerShellVerion -majorVersion 5
+
+if(!(Test-IsRunAsAdministrator))
```
[the remark this answers is not in the document]
Comment (hc-jordenlowe) [reply Fabric.Identity.API/scripts/Install-Identity-Discovery-IdPSS.ps1]: It was in the identity script before.  If it provides no value, then we can remove it.

File: Fabric.Identity.API/scripts/Install-Discovery.ps1 [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+$discoverySettingsScope = "discoveryservice"
+$discoveryConfig = Get-DosConfigValues -ConfigStore $configStore -Scope $discoverySettingsScope
+$commonConfig = Get-DosConfigValues -ConfigStore $configStore -Scope "common"
+$identityConfig = Get-DosConfigValues -ConfigStore $configStore -Scope "identity"
+Set-LoggingConfiguration -commonConfig $commonConfig
+
+# Pre-check requirements to run this script
+if([string]::IsNullOrEmpty($commonConfig.webServerDomain))
+{
+  Write-DosMessage -Level "Error" -Message "It is required to have 'webServerDomain' populated in install.config."
```
[the remark this answers is not in the document]
Comment (hc-jordenlowe) [reply Fabric.Identity.API/scripts/Install-Discovery.ps1]: A great, ill take a look.

File: Fabric.Identity.API/scripts/Install-Discovery.ps1 [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+}
+
+if([string]::IsNullOrEmpty($commonConfig.clientEnvironment))
+{
+  Write-DosMessage -Level "Error" -Message "It is required to have 'clientEnvironment' populated in install.config."
+}
+
+$discoveryIisUser = Get-IISAppPoolUser -credential $credential -appName $discoveryConfig.appName -storedIisUser $discoveryConfig.iisUser -installConfigPath $configStore.Path -scope $discoverySettingsScope
+
+$discoveryStandalonePath = ".\Catalyst.DiscoveryService.zip"
```
[the remark this answers is not in the document]
Comment (hc-jordenlowe) [reply Fabric.Identity.API/scripts/Install-Discovery.ps1]: This is the pattern from the other scripts.  This gets concatenated I believe.  Ill double check this.

File: Fabric.Identity.API/scripts/Install-Identity-Discovery-IdPSS.ps1 (hunk 7 above)
[the remark this answers is not in the document]
Comment (hc-jordenlowe) [reply Fabric.Identity.API/scripts/Install-Identity-Discovery-IdPSS.ps1]: Im going to keep it this way to smooth over some of this.  The other scripts can take in a config store.

File: Fabric.Identity.API/scripts/Install-Identity-Utilities.psm1 (hunk 6 above)
Comment (hc-jordenlowe) [reply Fabric.Identity.API/scripts/Install-Identity-Utilities.psm1]: That would require a huge amount of refactoring.  Im not disagreeing.  Perhaps we can add a PBI for a product enhancement.

File: Fabric.Identity.API/scripts/Install-Discovery.ps1 (hunk 9 above)
[the remark this answers is not in the document]
Comment (hc-jordenlowe) [reply Fabric.Identity.API/scripts/Install-Discovery.ps1]: Would throw work?

[the remark this answers is not in the document]
Comment (hc-jordenlowe) [reply Fabric.Identity.API/scripts/Install-Discovery.ps1]: ok i see later there is -ErrorAction Stop, so i just added that.

**Pull request merged**